File: fpaq0.java
import java.io.*;

public final class fpaq0 {
  public static final class State {
    public int x1;
    public int x2;
    public int p;
    public int cxt;
    public int[][] ct;
    public State(int x1, int x2, int p, int cxt, int[][] ct) {
      this.x1 = x1;
      this.x2 = x2;
      this.p = p;
      this.cxt = cxt;
      this.ct = ct;
    }
    public State() {
      this(0, 0xFFFFFFFF, 2048, 1, new int[256][2]);
    }
  }
  public static short predict(State s, int y) {
    if (++s.ct[s.cxt][y] > 65534)
      { s.ct[s.cxt][0] >>= 1; s.ct[s.cxt][1] >>= 1; }
    if ((s.cxt += s.cxt + y) >= 256) s.cxt = 1;
    return (short) (4096 * (s.ct[s.cxt][1] + 1) / (s.ct[s.cxt][0] + s.ct[s.cxt][1] + 2));
  }
  public static void ac_flush(State s, BufferedOutputStream out) throws IOException {
    out.write(s.x1 >>> 24);
  }
  public static void ac_rescale(State s) {
    s.x1 <<= 8;
    s.x2 = (s.x2 << 8) + 255;
  }
  public static void ac_encode_bit(State s, int y, BufferedOutputStream out) throws IOException {
    int range = s.x2 - s.x1;
    int xmid = s.x1 + (range >>> 12) * s.p + (((range & 0xfff) * s.p) >>> 12);
    s.p = predict(s, y);
    if (y == 1) s.x2 = xmid; else s.x1 = xmid + 1;
    while (((s.x1 ^ s.x2) & 0xff000000) == 0)
      { ac_flush(s, out); ac_rescale(s); }
  }
  public static void encode_file(State s, BufferedInputStream in, BufferedOutputStream out, long len) throws IOException {
    out.write((int) ((len >> 24) & 0xff));
    out.write((int) ((len >> 16) & 0xff));
    out.write((int) ((len >> 8) & 0xff));
    out.write((int) (len & 0xff));
    for (;;) {
      int c = in.read();
      if (c == -1) break;
      for (int i = 7; i >= 0; i--)
        ac_encode_bit(s, (c >> i) & 1, out);
    }
    ac_flush(s, out);
  }
  public static void main(String[] args) throws IOException {
    FileInputStream in = new FileInputStream(args[0]);
    FileOutputStream out = new FileOutputStream(args[1]);
    State main = new State();
    encode_file(main, new BufferedInputStream(in), new BufferedOutputStream(out), in.getChannel().size());
    in.close();
    out.close();
  }
}

File: fpaq0.js

const fs = require('fs');

let cxt = 1;  let ct = new Int32Array(256 * 2);
const predict = (y) => {
  if (++ct[cxt * 2 + y] > 65534)
    ct[cxt * 2] >>= 1, ct[cxt * 2 + 1] >>= 1;
  if ((cxt += cxt + y) >= 256) cxt = 1;
  return Math.floor((4096 * (ct[cxt * 2 + 1] + 1)) / (ct[cxt * 2] + ct[cxt * 2 + 1] + 2));
}

const BUFSIZ = 8192;

class InputWrapper {
  constructor(input) {
    this.input = input;
    this.buffer = new Uint8Array(BUFSIZ);
    this.pos = 8192;
  }
  read() {
    if (this.pos >= BUFSIZ) {
      fs.readSync(this.input, this.buffer, 0, BUFSIZ, null);
      this.pos = 0;
    }
    return this.buffer[this.pos++];
  }
}

class OutputWrapper {
  constructor(output) {
    this.output = output;
    this.buffer = new Uint8Array(BUFSIZ);
    this.pos = 0;
  }
  write(x) {
    if (this.pos >= BUFSIZ) {
      fs.writeSync(this.output, this.buffer, 0, this.pos, null);
      this.pos = 0;
    }
    this.buffer[this.pos++] = x;
  }
  flush() {
    fs.writeSync(this.output, this.buffer, 0, this.pos, null);
  }
}

class AC {
  constructor(stream) {
    this.x1 = 0;
    this.x2 = 0xFFFFFFFF;
    this.p = 2048;
    this.stream = stream;
  }
  ac_flush() {
    this.stream.write(this.x1 >> 24);
  }
  ac_rescale() {
    this.x1 <<= 8;
    this.x2 = (this.x2 << 8) + 0xFF;
  }
  ac_encode_bit(y) {
    let range = this.x2 - this.x1;
    let xmid = this.x1 + (range >>> 12) * this.p + ((range & 0xfff) * this.p >>> 12);
    this.p = predict(y);
    if (y) this.x2 = xmid;
    else this.x1 = xmid + 1;
    while (((this.x1 ^ this.x2) & 0xFF000000) === 0)
      { this.ac_flush(); this.ac_rescale(); }
  }
}

const encode = (input, output, len) => {
  let ac = new AC(output);
  output.write((len >> 24) & 0xFF);
  output.write((len >> 16) & 0xFF);
  output.write((len >> 8) & 0xFF);
  output.write(len & 0xFF);
  for (let i = 0; i < len; i++) {
    let x = input.read();
    for (let j = 7; j >= 0; j--)
      ac.ac_encode_bit((x >> j) & 1);
  }
  ac.ac_flush();
};

if (process.argv.length !== 4) {
  console.log("Usage: node fpaq0.js input output");
  process.exit(1);
}

const input = new InputWrapper(fs.openSync(process.argv[2], 'r'));
const output = new OutputWrapper(fs.openSync(process.argv[3], 'w'));
encode(input, output, fs.fstatSync(input.input).size);
output.flush();